### PR TITLE
[api] Mark SetStatus & GetStatus ActivityExtensions obsolete

### PR DESCRIPTION
**Pull request by ysolomchenko**
Fixes https://github.com/open-telemetry/opentelemetry-dotnet/issues/4755
Design discussion issue https://github.com/open-telemetry/opentelemetry-dotnet/discussions/4703

## Changes

Mark SetStatus activity extensions method obsolete

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry.Api/Trace/ActivityExtensions.cs [hunk 1]
```diff
@@ -25,9 +25,14 @@ public static class ActivityExtensions
     /// This extension provides a workaround to store Status as special tags with key name of otel.status_code and otel.status_description.
     /// Read more about SetStatus here https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#set-status.
     /// </summary>
+    /// <remarks>
```
Comment (lachmatt) [inline src/OpenTelemetry.Api/Trace/ActivityExtensions.cs:28]: I think summary should be cleanup as well, as suggested [here](https://github.com/open-telemetry/opentelemetry-dotnet/pull/4864#discussion_r1330430968).

File: src/OpenTelemetry.Api/Trace/ActivityExtensions.cs [hunk 2]
```diff
@@ -25,9 +25,14 @@ public static class ActivityExtensions
...
+    /// This method is obsolete.
+    /// For more details see: <see href="https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Api#setting-status" />.
+    /// </remarks>
     /// <param name="activity">Activity instance.</param>
     /// <param name="status">Activity execution status.</param>
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    [Obsolete("This method is obsolete. Use Activity.SetStatus(ActivityStatusCode) instead.")]
```
Comment (lachmatt) [inline src/OpenTelemetry.Api/Trace/ActivityExtensions.cs:35]:
```suggestion
    [Obsolete("Use Activity.SetStatus(ActivityStatusCode) instead.")]
```

File: src/OpenTelemetry.Api/Trace/ActivityExtensions.cs (hunk 1 above)
Comment (ysolomchenko) [reply src/OpenTelemetry.Api/Trace/ActivityExtensions.cs:25]: Done

File: src/OpenTelemetry.Api/Trace/ActivityExtensions.cs (hunk 2 above)
Comment (ysolomchenko) [reply src/OpenTelemetry.Api/Trace/ActivityExtensions.cs]: Done

File: src/OpenTelemetry.Api/Trace/ActivityExtensions.cs [hunk 3]
```diff
@@ -21,13 +21,15 @@ public static class ActivityExtensions
 {
     /// <summary>
     /// Sets the status of activity execution.
-    /// Activity class in .NET does not support 'Status'.
-    /// This extension provides a workaround to store Status as special tags with key name of otel.status_code and otel.status_description.
-    /// Read more about SetStatus here https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#set-status.
     /// </summary>
+    /// <remarks>
+    /// This method is obsolete.
```
Comment (cijothomas) [inline src/OpenTelemetry.Api/Trace/ActivityExtensions.cs:26]: can you give the alternate API to use right here, so users dont have to open the link and read about it? 
(The content from changelog.md file is good to be put here.)

File: src/OpenTelemetry.Api/Trace/TelemetrySpan.cs [hunk 4]
```diff
@@ -47,7 +47,11 @@ public ActivitySpanId ParentSpanId
     /// <param name="value">Status to be set.</param>
     public void SetStatus(Status value)
     {
+#pragma warning disable
+        // TODO: Remove this when SetStatus is deprecated
         this.Activity?.SetStatus(value);
+#pragma warning disable
+        this.Activity?.SetStatus((ActivityStatusCode)value.StatusCode);
```
Comment (cijothomas) [inline src/OpenTelemetry.Api/Trace/TelemetrySpan.cs:54]: not sure why we do this both ways?

File: src/OpenTelemetry/Trace/Processor/ExceptionProcessor.cs [hunk 5]
```diff
@@ -62,8 +62,10 @@ public override void OnEnd(Activity activity)
 
         if (snapshot != pointers)
         {
+#pragma warning disable
             // TODO: Remove this when SetStatus is deprecated
             activity.SetStatus(Status.Error);
```
Comment (cijothomas) [inline src/OpenTelemetry/Trace/Processor/ExceptionProcessor.cs:67]: could this be moved to the new SetStatus API?

File: src/OpenTelemetry.Api/Trace/TelemetrySpan.cs (hunk 4 above)
Comment (vishweshbankwar) [reply src/OpenTelemetry.Api/Trace/TelemetrySpan.cs:54]: this is needed for backward compatibility. if we simply remove the older way, we might break code/libraries still relying on older way.

File: src/OpenTelemetry/Trace/Processor/ExceptionProcessor.cs (hunk 5 above)
Comment (ysolomchenko) [reply src/OpenTelemetry/Trace/Processor/ExceptionProcessor.cs:67]: It already has a new SetStatus. As @vishweshbankwar mentioned, we need both ways.

File: src/OpenTelemetry.Api/Trace/ActivityExtensions.cs (hunk 3 above)
Comment (ysolomchenko) [reply src/OpenTelemetry.Api/Trace/ActivityExtensions.cs:26]: Done

**Comment by vishweshbankwar** [comment]
@ysolomchenko - Please take a look at test failures.

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpTraceExporterTests.cs [hunk 6]
```diff
@@ -370,7 +370,11 @@ public void ToOtlpSpanTest()
             links: childLinks);
 
         Assert.NotNull(childActivity);
+#pragma warning disable
+        // TODO: Remove this when SetStatus is deprecated
         childActivity.SetStatus(Status.Error);
+#pragma warning disable
+        childActivity.SetStatus(ActivityStatusCode.Error);
```
Comment (TimothyMothra) [inline test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpTraceExporterTests.cs:377]: I don't see where the status code is asserted in this test.
I see status description on line 397.

Just curious, if you call `SetStatus` and provide two different codes, which would win first or last?

File: src/OpenTelemetry.Api/CHANGELOG.md [hunk 7]
```diff
@@ -6,6 +6,12 @@ Notes](../../RELEASENOTES.md).
 
 ## Unreleased
 
+* Marked `ActivityExtensions.SetStatus` method as `[Obsolete]`.
+  Users are now encouraged to use the `Activity.SetStatus(ActivityStatusCode)`
```
Comment (vishweshbankwar) [inline src/OpenTelemetry.Api/CHANGELOG.md:10]:
```suggestion
  Users are now encouraged to use the [Activity.SetStatus(ActivityStatusCode](https://learn.microsoft.com/dotnet/api/system.diagnostics.activity.setstatus)
```

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpTraceExporterTests.cs (hunk 6 above)
Comment (vishweshbankwar) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpTraceExporterTests.cs]: There is a separate test for this:https://github.com/open-telemetry/opentelemetry-dotnet/blob/940d5b528dcfe532795f088a968b7cb13695f550/test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpTraceExporterTests.cs#L541

File: test/OpenTelemetry.Exporter.Zipkin.Tests/ZipkinExporterTests.cs [hunk 8]
```diff
@@ -554,7 +554,10 @@ public void UpdatesServiceNameFromIConfiguration()
 
         if (status.HasValue)
         {
+#pragma warning disable
             activity.SetStatus(status.Value);
+#pragma warning enable
+            activity.SetStatus((ActivityStatusCode)status.Value.StatusCode);
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Exporter.Zipkin.Tests/ZipkinExporterTests.cs:560]: @ysolomchenko - I think this assert is causing the issue in tests. I would just remove it for now. As a follow up, we can add additional test for validating the precedence like https://github.com/open-telemetry/opentelemetry-dotnet/pull/5781#discussion_r1717157416

File: src/OpenTelemetry.Api/CHANGELOG.md [hunk 9]
```diff
@@ -6,6 +6,12 @@ Notes](../../RELEASENOTES.md).
 
 ## Unreleased
 
+* Marked `ActivityExtensions.SetStatus` method as `[Obsolete]`.
+  Users are now encouraged to use the [Activity.SetStatus(ActivityStatusCode](https://learn.microsoft.com/dotnet/api/system.diagnostics.activity.setstatus)
```
Comment (vishweshbankwar) [inline src/OpenTelemetry.Api/CHANGELOG.md:10]:
```suggestion
  Users are now encouraged to use the [Activity.SetStatus(ActivityStatusCode)](https://learn.microsoft.com/dotnet/api/system.diagnostics.activity.setstatus)
```

**Review by vishweshbankwar — APPROVED** [review]
LGTM

**Review by vishweshbankwar — APPROVED** [review]

**Pull request merged**